Implement frame format

arthurprs opened the pull request:
Hello there :wave: I managed to get a mostly complete and efficient implementation for the frame format on top of your work. It'd be great if we could eventually reach a production level LZ4 implementation in Rust :tada: 

Please forgive the size of the PR, it's a bit larger than it should be. I may have sneaked a few improvements in this PR. While I was grokking the codebase I fixed/changed a few bits to be clearer/idiomatic.

It'd be great if you could take an initial look :smile: 

TODO:
- [x] Restore optimized `write_integer` implementation
- [x] Merge conflicts with master
- [x] Review safe compressor
- [x] Review safe de-compressor
- [X] Update unsafe de-compressor to work with frame format
- [x] Fuzzy tester for frame format
- [x] Benches for frame format
- [x] Review field names for buffer tracking in Frame (de)compressor
- [x] Add `enum CompressionError { OutputTooSmall }`
- [x] Make sure safe decompressor passes `fuzz_decomp_corrupted_data`
- [x] Export block stuff from block module and not its submodules.

PSeitz commented:
Nice work!
What I'm currently working on, and is probably the cause of the merge conflicts, is supporting slices as parameters instead of vecs (https://github.com/PSeitz/lz4_flex/issues/11). This is done to better support the frame format and also apply some performance optimizations. There are some implications which are not so easy to handle in the unsafe decompression, I'm currently looking into. It's probably better to use the new API.

Currently it accepts a `Sink` which is a u8 slice with pos. Not completely sure if this slice and pos should be wrapped or accepted as separate parameters in the API.

PSeitz commented on src/block/compress.rs:
```diff
@@ -124,83 +127,72 @@ fn token_from_literal_and_match_length(lit_len: usize, duplicate_length: usize)
             let diff = input_block ^ match_block;
-            num += get_common_bytes(diff) as usize;
-            break;
+            num += (diff.to_le().trailing_zeros() / 8) as usize;
```
`>> 3` should be faster than `/8`

arthurprs replied:
This helps quickly scanning the code. The performance aspect isn't applicable nowadays. Any decent compiler will emit >> 3 when it sees a division by 8. Rust will do that even in debug mode.

PSeitz commented on Cargo.toml:
```diff
@@ -18,9 +18,10 @@ path = "benches/crit_bench.rs"
 [dev-dependencies]
 criterion = "0.3.3"
-lz4 = "1.23.1"
+lzzzz = "0.8"
```
why did you change the lz4 to lzzzz?

arthurprs replied:
lzzzz supports all lz4(C) features, including the frame format. I need that to have a baseline implementation to compare against.

PSeitz commented on src/block/compress.rs:
```diff
@@ -124,83 +127,72 @@ fn token_from_literal_and_match_length(lit_len: usize, duplicate_length: usize)
+/// `candidate` is the candidate position in `source`
 ///
-/// The function ignores the last 7bytes (END_OFFSET) in input as this should be literals.
+/// The function ignores the last X bytes (END_OFFSET) in input as this should be literals.
```
last END_OFFSET bytes

arthurprs commented:
Honestly, after writing this changeset I don't think supporting a Vec for output internally is worth the effort. The API level is unaffected as we can just create a properly sized Vec internally and use the backing slice internally.
I agree this may have some slight performance implications in the unchecked decompressor but that's extremely unsafe.

But I really don't agree that a Vec helps in the frame format. The frame format enforces a strict max block size, which can be used to determine the max output size for the compressor and the the size of the output slice for the decompressor.

arthurprs commented:
For example, consider this snippet from the safe decompressor

```
if output_len + 24 < output.len() {
    if match_length <= 16 {
        output.copy_within(start..start + 16, output_len);
    } else {
        // TODO: mark cold
        output.copy_within(start..start + 24, output_len);
    }
} else {
    output.copy_within(start..start + match_length, output_len)
}
```

There's a very predictable branch to enable the wild copy, equivalent to the branch that would be used to check capacity in a Vec.

arthurprs commented:
I fixed the conflicts with master. I think the compressor is in a good state, if you want to start a closer review. I'll move onto the safe decompressor when time allows.

PSeitz commented on src/block/compress.rs:
```diff
@@ -252,69 +242,61 @@ fn count_same_bytes(input: &[u8], candidate: usize, cur: &mut usize) -> usize {
-        core::ptr::copy_nonoverlapping(el.to_le_bytes().as_ptr(), out_ptr, 4);
-    }
+    // Write the last 0..=3 0xFF bytes + remainder.
+    let ffs = n / 0xFF;
```
what does ffs mean?

arthurprs replied:
It's meant to be the number of `0xFF`s to be output. Sorry :facepalm: quite bad naming from my part.

PSeitz commented on src/block/compress.rs:
```diff
@@ -252,69 +242,61 @@ fn count_same_bytes(input: &[u8], candidate: usize, cur: &mut usize) -> usize {
+    source: &[u8],
+    candidate: &mut usize,
+) {
+    let left = input[literal_start..*cur].iter().rev().copied();
+    let right = source[..*candidate].iter().rev().copied();
+    for (a, b) in left.zip(right) {
+        if a != b {
+            break;
+        }
```
Did you benchmark this? I tried a similar solution which was slower

arthurprs replied:
I think I did, but I'll double check :+1:

PSeitz commented on src/block/compress.rs:
```diff
@@ -125,88 +128,72 @@ fn token_from_literal_and_match_length(lit_len: usize, duplicate_length: usize)
         }
     }
 
+    num += cur_slice
+        .iter()
+        .zip(cand_slice)
+        .skip(num)
+        .take_while(|(a, b)| a == b)
+        .count();
```
Isn't this covered by the loop above? This seems to cause a performance regression

arthurprs replied:
Unfortunately it's not, the block loop may break early in the end of input/source. This was already being handled in the unsafe version of the function.

PSeitz replied:
Since this barely affects compression ratio, I would take the tradeoff for more performance

PSeitz commented on src/block/compress.rs:
```diff
@@ -125,88 +128,72 @@ fn token_from_literal_and_match_length(lit_len: usize, duplicate_length: usize)
             let diff = input_block ^ match_block;
-            num += get_common_bytes(diff) as usize;
-            break;
+            num += (diff.to_le().trailing_zeros() / 8) as usize;
```
I added the `get_common_bytes` function mainly to increase readability, to understand what this code is doing

arthurprs replied:
I inlined to make it similar to the other locations dealing with u32/u16 instead (in the unsafe version of this function).

Zero extending big/little-endian u32/u16 to usize to feed to get_common_bytes is very error prone when we're counting zero bits. I think this way strikes a good balance of explicit/readability/uniformity. What do you think?

PSeitz replied:
I think it's fine, but a reader not familiar with the code may not be able to understand semantically what the code does, to count the common bytes

PSeitz reviewed (commented):
I did run the safe compression benchmarks and there seems to be a performance regression (20-40% slower)
One reason seems to be an additional (superfluous?) count block in `count_same_bytes`.
Not sure if it's also related to the dictionary. If that's the case, I don't think the dictionary should cause performance issues since it's usage is probably not common.

arthurprs commented on src/block/compress.rs:
```diff
@@ -528,20 +564,39 @@ pub fn get_maximum_output_size(input_len: usize) -> usize {
+    while i + core::mem::size_of::<usize>() <= dict_data.len() {
+        let hash = get_hash_at(dict_data, i);
+        dict.put_at(hash, i);
+        i += 3;
```
I'm not sure, I copied it from the C implementation. It's probably some "sweet-spot" amount that the original author found.
Probably worth adding a comment?

PSeitz replied:
yes, a comment would be good.

There is a similar case I copied from the C implementation, which probably also should get a comment
https://github.com/PSeitz/lz4_flex/blob/master/src/block/compress.rs#L434-L435

arthurprs commented:
I can reproduce some regression with LTO enabled, but much smaller ones :thinking:  I suspect there are differences in the benchmarks so I added a lz4_flex_master variant to the benchmarks (ed6e060) so they can be run alongside.

**W/O LTO**
faster (see LTO bellow)
```
Compress/lz4_flexx_rust/725                                                                             
                        time:   [1.2502 us 1.2516 us 1.2529 us]
                        thrpt:  [551.83 MiB/s 552.44 MiB/s 553.05 MiB/s]
                 change:
                        time:   [-1.7243% -1.5582% -1.3910%] (p = 0.00 < 0.05)
                        thrpt:  [+1.4106% +1.5829% +1.7545%]
                        Performance has improved.
Found 1 outliers among 100 measurements (1.00%)
  1 (1.00%) high severe
Compress/lz4_flexx_rust_master/725                                                                             
                        time:   [1.7235 us 1.7285 us 1.7336 us]
                        thrpt:  [398.83 MiB/s 400.01 MiB/s 401.18 MiB/s]
Found 5 outliers among 100 measurements (5.00%)
  3 (3.00%) high mild
  2 (2.00%) high severe
Compress/lz4_flexx_rust/34308                                                                            
                        time:   [92.714 us 93.004 us 93.467 us]
                        thrpt:  [350.06 MiB/s 351.80 MiB/s 352.90 MiB/s]
                 change:
                        time:   [-2.1264% -1.2183% +0.0655%] (p = 0.01 < 0.05)
                        thrpt:  [-0.0654% +1.2333% +2.1726%]
                        Change within noise threshold.
Found 9 outliers among 100 measurements (9.00%)
  3 (3.00%) high mild
  6 (6.00%) high severe
Compress/lz4_flexx_rust_master/34308                                                                            
                        time:   [101.93 us 102.23 us 102.68 us]
                        thrpt:  [318.65 MiB/s 320.04 MiB/s 320.98 MiB/s]
Found 11 outliers among 100 measurements (11.00%)
  1 (1.00%) high mild
  10 (10.00%) high severe
  ```


**W/ WTO**
10% regression for 725 bytes
4% regression for 32KB
The massive improvement from non-lto to lto for master suggests it was missing some inline annotations. I added a few in this PR.

```
Compress/lz4_flexx_rust/725                                                                             
                        time:   [1.2312 us 1.2455 us 1.2695 us]
                        thrpt:  [544.66 MiB/s 555.15 MiB/s 561.58 MiB/s]
                 change:
                        time:   [-1.7319% -1.1970% -0.5335%] (p = 0.00 < 0.05)
                        thrpt:  [+0.5363% +1.2115% +1.7624%]
                        Change within noise threshold.
Found 12 outliers among 100 measurements (12.00%)
  6 (6.00%) high mild
  6 (6.00%) high severe
Compress/lz4_flexx_rust_master/725                                                                             
                        time:   [1.1240 us 1.1265 us 1.1298 us]
                        thrpt:  [611.99 MiB/s 613.75 MiB/s 615.13 MiB/s]
                 change:
                        time:   [-35.620% -35.148% -34.771%] (p = 0.00 < 0.05)
                        thrpt:  [+53.307% +54.197% +55.327%]
                        Performance has improved.
Found 7 outliers among 100 measurements (7.00%)
  1 (1.00%) low mild
  2 (2.00%) high mild
  4 (4.00%) high severe
Compress/lz4_flexx_rust/34308                                                                            
                        time:   [90.973 us 91.036 us 91.102 us]
                        thrpt:  [359.14 MiB/s 359.40 MiB/s 359.65 MiB/s]
                 change:
                        time:   [-3.4910% -2.4520% -1.7169%] (p = 0.00 < 0.05)
                        thrpt:  [+1.7469% +2.5136% +3.6173%]
                        Performance has improved.
Found 4 outliers among 100 measurements (4.00%)
  1 (1.00%) high mild
  3 (3.00%) high severe
Compress/lz4_flexx_rust_master/34308                                                                            
                        time:   [87.705 us 87.840 us 87.979 us]
                        thrpt:  [371.89 MiB/s 372.48 MiB/s 373.05 MiB/s]
                 change:
                        time:   [-16.416% -15.292% -14.409%] (p = 0.00 < 0.05)
                        thrpt:  [+16.834% +18.052% +19.640%]
                        Performance has improved.
Found 7 outliers among 100 measurements (7.00%)
  2 (2.00%) low mild
  2 (2.00%) high mild
  3 (3.00%) high severe
```

PSeitz commented:
Ah sorry I changed Cargo.toml on master to use unsafe by default from the last dev session, the comparison is invalid

PSeitz commented:
Looks really good so far, nice job!
Would be good to add some benchmarks with the frame format. It would also good to enable `COMPRESSION66K`, this is a JSON data example, which has often different performance characteristics.

PSeitz commented on src/block/decompress_safe.rs:
```diff
@@ -126,38 +121,47 @@ pub fn decompress_into(input: &[u8], output: &mut Sink) -> Result<(), Decompress
-            // The blocks can overlap, make sure they are at least BLOCK_COPY_SIZE apart
-            if match_length + BLOCK_COPY_SIZE >= offset {
+            // The blocks can overlap, make sure they are at least 20 bytes apart
+            if offset > output.pos() {
+                let copied = copy_from_dict(output, ext_dict, offset, match_length)?;
+                if copied == match_length {
+                    continue;
+                }
+                // match crosses ext_dict and output, offset is still correct as output_len was increased
+                match_length -= copied;
+            }
```
The dictionary support here causes a performance regression of around 10%

arthurprs commented:
I was able to reduce all sources of overhead and the non-dict version is now faster across the board.
![image](https://user-images.githubusercontent.com/715958/114271970-5621a100-9a14-11eb-939a-3f5832320be7.png)

![image](https://user-images.githubusercontent.com/715958/114271967-51f58380-9a14-11eb-968f-65a9fc1e5bb8.png)

PSeitz commented:
Nice usage of const generics!

PSeitz commented:
I added some comments directly on the commits, I hope that's fine

billyb2 commented:
Just out of curiosity, how would I use the frame decompression? Could you please give an example? Thank you 😄

arthurprs commented:
@billyb2 here're examples, it's likely that similar functions will be added to the crate.

```rust
fn compress_frame(input: &[u8]) -> Vec<u8> {
    let mut enc = FrameEncoder::new(Vec::new());
    enc.write_all(input).unwrap();
    enc.finish().unwrap()
}

fn decompress_frame(input: &[u8]) -> Result<Vec<u8>, lz4_flex::frame::Error> {
    let mut de = FrameDecoder::new(input);
    let mut out = Vec::new();
    de.read_to_end(&mut out)?;
    Ok(out)
}
```

PSeitz commented on src/block/decompress.rs:
```diff
@@ -276,15 +278,15 @@ fn decompress_internal<const USE_DICT: bool>(
             // In this branch we know that match_length is at most 18 (14 + MINMATCH).
-            // But the blocks can overlap, so make sure they are at least 20 bytes apart
-            // to enable an optimized non-overlaping copy of 20 bytes.
-            if offset < 20 {
```
I chose here 20, because it's dividable by 4, but I didn't actually run benchmarks between 18 and 20 for it. Just FYI

arthurprs replied:
I think it was 16 w/ a branch to 24 before. I got improvements switching to a fixed 20, but fixed 18 should be equaliy sensible.

arthurprs commented:
I think the coverage setup is not working as expected. Specifically I don't think it's merging the results of the multiple runs (w/ different features).

PSeitz commented:
> I think the coverage setup is not working as expected. Specifically, I don't think it's merging the results of the multiple runs (w/ different features).

Yes it doesn't correctly aggregate runs with different feature flags, you can ignore it

arthurprs commented:
@PSeitz Thank you for you patience. I think it's ready for review. I added as much testing as I could so I'm cautiously optimistic it's working.

PSeitz commented:
@arthurprs thank you, really impressive job there. reviewing all of it may take a little.

PSeitz commented on src/block/decompress.rs:
```diff
@@ -23,11 +23,29 @@ unsafe fn duplicate(
+#[inline]
+fn wild_copy_from_src_16(mut source: *const u8, mut dst_ptr: *mut u8, num_items: usize) {
+    // Note: if the compiler transforms this into a call to memcpy it'll hurt performance!
+    // It doesn't seem to be the case though, probably due to the 16 byte stride length.
```
really good catch, I also noticed that 16bytes was much faster but thought it was because of the larger register.
I wonder if this is worth reporting as compiler a bug

arthurprs replied:
Yeah, misoptimizations like this are tricky. This is likely an intended optimization in LLVM. I'm actually surprised the 16byte loop isn't also being "optimized".

PSeitz replied:
I reported it here: https://github.com/rust-lang/rust/issues/84813

PSeitz commented on src/frame/mod.rs:
```diff
@@ -1,115 +1,82 @@
+            + crate::block::compress::get_maximum_output_size(input.len()),
+    );
+    let mut enc = FrameEncoder::with_frame_info(frame_info, buffer);
+    enc.write_all(input).unwrap();
+    enc.finish().unwrap()
```
I think the errors should be forwarded here

arthurprs replied:
I wasn't sure what to do here. In theory this will always succeed, as reading from bytes and writing to a vec always work.

PSeitz commented on src/frame/header.rs:
```diff
@@ -0,0 +1,333 @@
+*/
+#[derive(Debug, Clone)]
+pub struct FrameInfo {
+    pub content_size: Option<u64>,
+    pub dict_id: Option<u32>,
+    pub block_size: BlockSize,
+    pub block_mode: BlockMode,
+    pub block_checksums: bool,
+    pub content_checksum: bool,
```
these should be documented

currently there is no `#![warn(missing_docs)]` in the crate root, but it should probably be added to make sure everything is documented

arthurprs replied:
Good point, I'll get on it.

PSeitz commented on src/frame/decompress.rs:
```diff
@@ -1 +1,312 @@
+/// This reader can potentially make many small reads from the underlying
+/// stream depending on its format, therefore, passing in a buffered reader
+/// may be beneficial.
+pub struct FrameDecoder<R: io::Read> {
```
Is there some advantage exposing `FrameDecoder` instead of having something like

``` pub fn decompress_into(input: &[u8], out: impl Writer) -> Result<(), Error> {```

arthurprs replied:
Yes, having `FrameDecoder` allow the user to wrap any `io::Read` and use that anywhere that takes an `io::Read`.

PSeitz replied:
I mean from an API it would be simpler to have
`pub fn decompress_into(input: impl Read, out: impl Write) -> Result<(), Error> {`

does `FrameDecoder` add some functionality we need or can we keep it private?

arthurprs replied:
It doesn't expose additional functionality but it's _existence_ is useful because it implements `io::Read`. For instance, you can use it on a function like https://docs.serde.rs/serde_json/fn.from_reader.html

PSeitz replied:
I'm thinking about the _public_ API. Locally I added this to frame/mod.rs, This works fine and has a simple API. Do we get additional functionality by exposing `FrameDecoder`?

```

/// Decompress all bytes of `input` into a new vec.
pub fn decompress_into(input: impl Read, mut out: impl Write) -> Result<(), Error> {
    let mut de = FrameDecoder::new(input);
    io::copy(&mut de, &mut out)?;
    Ok(())
}

/// Compress all bytes of `input`.
pub fn compress_with_into(
    frame_info: FrameInfo,
    mut input: impl Read,
    out: impl Write,
) -> Result<u64, std::io::Error> {
    let mut enc = FrameEncoder::with_frame_info(frame_info, out);
    let bytes_written = io::copy(&mut input, &mut enc)?;
    enc.finish()?; // todo missing last bytes
    Ok(bytes_written)
}

#[test]
fn test_frame() {
    use std::fs;
    let input = fs::File::open("./benches/compression_66k_JSON.txt").unwrap();
    let output = fs::File::create("./tmp.file").unwrap();
    compress_with_into(FrameInfo::default(), input, output).unwrap();
    let input = fs::File::open("./tmp.file").unwrap();
    let restore = fs::File::create("./tmp.restore.file").unwrap();
    decompress_into(input, restore).unwrap();
    let restore_data = fs::read("./tmp.restore.file").unwrap();
    let original = fs::read("./benches/compression_66k_JSON.txt").unwrap();
    assert_eq!(restore_data, original);
}
```

arthurprs replied:
Yes. Like in my example, you want use it as an io::Read directly.

I've used this many times and it was the motivator for me to write this in the first place.

```rust
fn test() {
    let compressed_input = std::fs::File::open("datafile").unwrap(); // or any other io::Read like a &[u8]
    let mut decompressed_input = lz4_flex::frame::FrameDecoder(compressed_input).unwrap();
    let json: serde_json::Value = serde_json::from_reader::<serde_json::Value>(decompressed_input).unwrap();

    // or if you have many json values in the input
    let compressed_input = std::fs::File::open("datafile").unwrap(); // or any other io::Read like a &[u8]
    let mut decompressed_input = lz4_flex::frame::FrameDecoder(compressed_input).unwrap();
    loop {
        match serde_json::from_reader::<serde_json::Value>(decompressed_input) {
            Ok(json) => { println!("json {:?}", json); }
            Err(e) if e.is_eof() => break,
            Err(e) => panic!("{}", e),
        }
        
    }
}
```

This is a read example, but the write side can use an FrameEncoder as well.

PSeitz replied:
Okay I get it, FrameDecoder is better to connect multiple streams. 

I think this is also a nice example to put into the docs

arthurprs replied:
Added examples to FrameEncoder/Decoder

PSeitz commented on src/frame/mod.rs:
```diff
@@ -1,115 +1,82 @@
-        1 << (self.block_size_id as u32)
-    }
+/// Decompress all bytes of `input` into a new vec.
+pub fn decompress(input: &[u8]) -> Result<Vec<u8>, Error> {
```
data can be very large, so we shouldn't require it to be in memory and accept `impl Read` instead

```pub fn decompress(input: &[u8]) -> Result<Vec<u8>, Error> {```

not sure about the api implications downstream here though..

arthurprs replied:
This is supposed to be an utility function to mirror the ones in the block module, in case the user wants to work with bytes/Vec.
Users wanting to interface with io::{Read, Write} are better of with the equivalent FrameEncoder/Decoder.

PSeitz replied:
The decision for the block API to support only in memory data is purely for performance reasons, `Read` and `Write` as an API would be better

arthurprs replied:
Indeed :thinking:  These utility functions should probably take io::{Read, Write} whenever possible.

arthurprs replied:
There's benefit in having easily reachable versions for slice/Vec as they doesn't require using `io::copy` (which involves an intermediary buffer) and also allow better preallocation.

For that reason I added `compress_into`, `compress_into_with` and `decompress_into` which take `io::{Read,Write}` for input/output respectively. It's not worth providing all combinations (eg. input is a slice output is an io::Write), so this probably strikes a good trade-off.

What do you think?

PSeitz replied:
I think this depends on the performance gain, you can also preallocate the vec and pass it as the `impl Write`

arthurprs replied:
Right, a cautions caller could preallocate, which leaves only io::copy overhead.
Don't you think this is the most common usage, so it warrants the optimized functions?

PSeitz replied:
since a Vec implements `Write` the usage would be same, so there needs to be some performance gain to justify the method

arthurprs replied:
I ran **compression** benchmarks and both the io::copy and not preallocating overheads are negligible for small data (<=64kb). Probably because compression is much slower than memory copies.

I only measured significant overhead with the 10mb dataset. 2% overhead for the io::copy and 2% more if the Vec is not preallocated.

For **decompression** I measured 3% overhead by not preallocating the output.

Surprisingly (to me at least) the io::copy version (decompress_into) was initially a whopping 18% faster than the specialized version (decompress) which uses read_to_end. This is because the default io::Read::read_to_end implementation is really suboptimal in this case.

Fixing FrameDecoder::read_to_end makes it 5% faster than the equivalent io::copy version.

At least we know no version is awfully worse than the other, and than pre-allocation doesn't matter much except for large data.

arthurprs replied:
With that said I think the current functions in frame/mod.rs are reasonable as they strike a good balance between API surface, ergonomics and optimization.

PSeitz replied:
Thank you for the insight. With gains of 2-5%, I'm not really convinced for the reduced API ergonomics. With frames, the default case is to stream them I think. In most cases, e.g. a web server, allocating 10MB for a decompression could already be too much. If you have always smaller data, you can use the faster block API.

I had a look at the snap crate: https://docs.rs/snap/1.0.5/snap/index.html

1. There are no such utility functions
2. There is `write::FrameEncoder` `read::FrameDecoder` and (surprisingly)`read::FrameEncoder ` 

I think the api of the utility function is easier to grasp, than `FrameEncoder` and `FrameDecoder`, but maybe that's just me and this can also be completely covered by documentation. 
Generally they are not necessary at all. So my feeling is, it's better to have a minimal, powerful API like the snap crate and document it well for all cases.

arthurprs replied:
Seems sensible. I updated the PR as such.

PSeitz commented on src/frame/decompress.rs:
```diff
@@ -1 +1,312 @@
+            ext_dict_len: 0,
+            dst_start: 0,
+            dst_end: 0,
+            frame_info: None,
```
I think it would be better read_frame_info in the constructor and make frame_info non optional

arthurprs replied:
I think it's a good practice that io::Read/Write wrappers are lazy so they don't do much unless read/write is called. 
The FrameInfo being present/not-present is important so an FrameDecoder can read multiple frames from the underlying io::Read (like in tests.rs `frame::concatenated`).

Frame info doesn't necessarily have to be an Option I guess. I could be set to the default value (and then read from the io::Read) and then we add a boolean or something to track if it's between a frame start and frame end. That's probably less intuitive though.

PSeitz replied:
I think it's okay to have FrameInfo as Option to have lazy readers, but the information that we are at the end is stored implicitly in the frame info option, I think it's always better to be explicit.

arthurprs replied:
Maybe renaming the field to `current_frame_info` helps?

PSeitz replied:
yes I think that's better

PSeitz commented on src/block/compress.rs:
```diff
@@ -324,117 +327,177 @@ pub fn backtrack_match(candidate: &mut usize, cur: &mut usize, literal_start: us
+/// Compress all bytes of `input[input_pos..]` into `output`.
+/// Bytes in `input[..input_pos]` are treated as a preamble and can be used for lookback.
+/// Bytes in `ext_dict` logically precede `pre[..input_pos]` and can also be used for lookback.
+/// `input_stream_offset` is the logical position of the first byte of `input` in the complete
```
having a block here explaining the difference between input_pos and input_stream_offset would be good

arthurprs replied:
I added a few lines, but TBH it is quite hard to explain.

PSeitz commented on src/block/compress.rs:
```diff
@@ -379,10 +383,11 @@ pub(crate) fn compress_internal<T: HashTable, const USE_DICT: bool>(
             .checked_add(input.len())
             .and_then(|i| i.checked_add(ext_dict.len()))
-            .map_or(false, |i| i <= usize::MAX / 2));
+            .map_or(false, |i| i <= usize::MAX));
```
`i <= usize::MAX`
this evaluates always to true

arthurprs replied:
Good point, I shouldn't have changed this.

PSeitz commented on src/frame/mod.rs:
```diff
@@ -54,29 +75,53 @@ impl fmt::Display for Error {
+/// Compress `input` into `output` with the specified Frame configuration.
+pub fn compress_into_with(
+    frame_info: FrameInfo,
+    input: &mut impl Read,
+    output: &mut impl Write,
```
Isn't it common to take `Read` and `Write` by value and the caller needs to make sure that the &mut references implement Write and Read

arthurprs replied:
I don't know honestly. I was following `io::copy` signature here, which is semantically equivalent.

```rust
pub fn copy<R: ?Sized, W: ?Sized>(reader: &mut R, writer: &mut W) -> Result<u64> where
    R: Read,
    W: Write,
```

arthurprs replied:
Having played with these functions a bit more I think forcing &mut onto the users (like io::copy) is quite unergonomic. Should I switch it like you suggested?

PSeitz replied:
Yes, I checked some API's and this seems to be the common case

PSeitz commented on src/frame/mod.rs:
```diff
@@ -54,29 +75,53 @@ impl fmt::Display for Error {
+    // Preallocate the output with 2x the input size, it may resize but it amortizes enough.
     // The upside is that we don't have to worry about DOS attacks, etc..
-    let mut out = Vec::with_capacity(input.len() * 3 / 2);
+    let mut out = Vec::with_capacity(input.len() * 2);
```
I think the compression can't be that bad? The maximum overhead should be much lower

arthurprs replied:
I'm not sure I understand what you mean. It's assuming a 0.5 compression ratio. Which is reasonable, right?

PSeitz replied:
ah I was thinking this preallocates for compression.
In the decompression case I don't know how this can affect DOS. I think for preallocation we should try to read the content size

arthurprs replied:
That's how you DOS. Build a valid frame with a massive content size.

arthurprs replied:
I considered doing that with a maximum limit to the allocation, but it required the specialisation feature. So when the  Frame Decoder read the FrameInfo it would try preallocating the output (with the specialisation trait).

Then I considered injecting an already read FrameInfo which seemed more doable but still significant tinkering with FrameDecoder (as it would not be reading the FrameInfo itself).

I ended up giving up and opting for the estimation based on the input size.

arthurprs replied:
What I was trying to say is that reading content length is very prone to DOS a naive decoder will preallocate `content_size`  bytes after only reading 15 bytes.

> If the frame is massive we will keep reading until the system runs out of Ram/disk space etc..
I see your point. It's still possible if the attacker can provide enough input, this one is not worth worrying about though.

PSeitz replied:
> 
> I see your point. It's still possible if the attacker can provide enough input, this one is not worth worrying about though.

With lz4 you can have a maximum ratio of around 250. Still a valid attack vector, but yes it's more difficult.

PSeitz commented on src/frame/header.rs:
```diff
@@ -0,0 +1,342 @@
+    /// The compressor and the decompressor must use exactly the same dictionary.
+    ///
+    /// Note that this is currently unsupported and if set will cause de (De)Compressor to error.
+    pub dict_id: Option<u32>,
```
I think this should not be pub, as long as it is not supported

arthurprs replied:
That's a good idea. Then the FrameEncoder construction can be infallible.

PSeitz commented on src/frame/header.rs:
```diff
@@ -0,0 +1,342 @@
+
+impl FrameInfo {
+    pub(crate) fn read_size(input: &[u8]) -> Result<usize, Error> {
+        let mut required = 7;
```
is the 7 MIN_FRAME_INFO_SIZE?

`let mut required = MIN_FRAME_INFO_SIZE;`

from the variable name, it's not clear what is required

arthurprs replied:
Yes, good catch.

PSeitz commented on src/frame/header.rs:
```diff
@@ -0,0 +1,345 @@
+        }
+        let mut hasher = XxHash32::with_seed(0);
+        hasher.write(&buffer[4..offset]);
+        let checksum = (hasher.finish() >> 8) as u8;
```
`let header_checksum` to distinguish from the other checksums

PSeitz commented on src/frame/decompress.rs:
```diff
@@ -19,6 +19,30 @@ use crate::block::WINDOW_SIZE;
 /// This reader can potentially make many small reads from the underlying
 /// stream depending on its format, therefore, passing in a buffered reader
 /// may be beneficial.
```
We can internally always wrap the `Read` with a `BufReader`. Forgetting to wrap something in a `BufReader` is a common performance issue I think

arthurprs replied:
TBH I copied this comment from the snap crate so it might very well be wrong as there's only 2 reads per block. 1 for the block header and another one to get its full contents.

Either way it's only really applicable to a file so we shouldn't force the overhead onto correct usages.

arthurprs replied:
I got these
```
FrameDecompress/lz4_flex_rust_buffered/725                                                                             
                        time:   [2.9677 us 2.9788 us 2.9905 us]
                        thrpt:  [231.20 MiB/s 232.11 MiB/s 232.98 MiB/s]
Found 16 outliers among 100 measurements (16.00%)
  4 (4.00%) low mild
  1 (1.00%) high mild
  11 (11.00%) high severe
FrameDecompress/lz4_flex_rust_raw/725                                                                             
                        time:   [3.8128 us 3.8289 us 3.8466 us]
                        thrpt:  [179.75 MiB/s 180.58 MiB/s 181.34 MiB/s]
Found 13 outliers among 100 measurements (13.00%)
  3 (3.00%) low mild
  7 (7.00%) high mild
  3 (3.00%) high severe
FrameDecompress/lz4_flex_rust_buffered/34308                                                                             
                        time:   [20.215 us 20.423 us 20.646 us]
                        thrpt:  [1.5476 GiB/s 1.5645 GiB/s 1.5806 GiB/s]
Found 10 outliers among 100 measurements (10.00%)
  6 (6.00%) high mild
  4 (4.00%) high severe
FrameDecompress/lz4_flex_rust_raw/34308                                                                             
                        time:   [20.474 us 20.750 us 21.049 us]
                        thrpt:  [1.5180 GiB/s 1.5398 GiB/s 1.5606 GiB/s]
Found 6 outliers among 100 measurements (6.00%)
  4 (4.00%) high mild
  2 (2.00%) high severe
FrameDecompress/lz4_flex_rust_buffered/66675                                                                             
                        time:   [21.209 us 21.547 us 21.939 us]
                        thrpt:  [2.8304 GiB/s 2.8819 GiB/s 2.9278 GiB/s]
Found 4 outliers among 100 measurements (4.00%)
  3 (3.00%) high mild
  1 (1.00%) high severe
FrameDecompress/lz4_flex_rust_raw/66675                                                                             
                        time:   [21.430 us 21.636 us 21.872 us]
                        thrpt:  [2.8390 GiB/s 2.8700 GiB/s 2.8977 GiB/s]
Found 4 outliers among 100 measurements (4.00%)
  4 (4.00%) high mild
```

So it only makes a difference for the tiny input. But the difference is still less than a microsecond, which is nothing when dealing with real files. So maybe that comment should be removed.

arthurprs replied:
Pushed a6fac63

PSeitz commented on fuzz/fuzz_targets/fuzz_decomp_corrupt_frame.rs:
```diff
@@ -0,0 +1,24 @@
+    other.clear();
+    other.extend_from_slice(&[0x04u8, 0x22, 0x4d, 0x18]);
+    other.extend_from_slice(data);
+    let _ = lz4_flex::frame::decompress(&other);
```
why do you bind against `let _ = `?

PSeitz commented on src/frame/compress.rs:
````diff
@@ -1,66 +1,361 @@
+/// ```
+///
+/// # Example 2
+/// Deserializing multiple json values out of a compressed file
````
the example code below is serialization. the examples are very similiar, was this intended?

arthurprs replied:
Yes they are intentionally similar. I think the second one can be improved to customizer the frame option, or somethinglike that.

PSeitz commented on src/frame/decompress.rs:
```diff
@@ -1 +1,387 @@
+            // Fill read buffer if there's uncompressed data left
+            if self.dst_start < self.dst_end {
+                let len = std::cmp::min(self.dst_end - self.dst_start, buf.len());
+                let dste = self.dst_start.checked_add(len).unwrap();
```
checked_add with unwrap is the same as just adding the numbers?

PSeitz commented on src/frame/decompress.rs:
```diff
@@ -1 +1,387 @@
+    /// that are still part of the decompressor window.
+    ext_dict_offset: usize,
+    ext_dict_len: usize,
+    /// Index into dst: starting point of bytes not yet given back to caller.
```
/// Index into dst: starting point of bytes not yet read by caller.

PSeitz commented on src/frame/decompress.rs:
```diff
@@ -1 +1,387 @@
+                let delta = self.ext_dict_len.min(self.dst_start);
+                self.ext_dict_offset += delta;
+                self.ext_dict_len -= delta;
+            }
```
what happens in the else case?

arthurprs replied:
This is complicated. I added a lot comments, copying the one I added to that block bellow.

>  There's more than WINDOW_SIZE of lookback adding the prefix and ext_dict.
>  Since we have a limited buffer we must shrink ext_dict in favor of the prefix,
>   so that we can fit up to max_block_size bytes between dst_start and ext_dict start.

PSeitz replied:
thanks they are really helpful

PSeitz commented on src/frame/decompress.rs:
```diff
@@ -1 +1,403 @@
+        {
+            self.src.clear();
+            self.dst.clear();
+            self.src.reserve_exact(max_block_size);
+            self.dst.reserve_exact(dst_size);
```
reserve_exact reserves additional capacity, not total. I think the intention here was the same as in resize to ensure the total capacity?

Although this only happens when we read concatenates frames, do you know if this is common?

arthurprs replied:
Yes, the sequence of clear->reserve_exact->set_len is meant to be a faster/uninitiated version of resize(). Since they are cleared just before the reserve_exact call it behaves as total capacity to ensure set_len will be valid.

This will happen on all frames, including single frame scenarios, as the buffers sizes ultimately depend on the fields of the FrameInfo.

My guess would be that concatenated frames are relatively rare.

PSeitz commented on src/frame/decompress.rs:
```diff
@@ -1 +1,403 @@
+        let frame_info = FrameInfo::read(&buffer[..required])?;
+        if frame_info.dict_id.is_some() {
+            // Unsupported right now so it must be None
+            return Err(Error::WrongDictionary {
+                actual: frame_info.dict_id,
+                expected: None,
+            }
```
I guess this feature is barely used anyway, but if someone runs into this, the error would be confusing

PSeitz replied:
A Unsupported or NotImplemented Error would be better I think

arthurprs replied:
The only drawback is that we can't remove the error variant later without a major version bump. I guess that's OK?

arthurprs commented on src/frame/compress.rs:
```diff
@@ -1,66 +1,364 @@
+
+        // Reposition the compression table if we're anywhere near an overflowing hazard
+        if self.src_stream_offset + max_block_size + WINDOW_SIZE >= u32::MAX as usize / 2 {
+            self.compression_table
+                .reposition((self.src_stream_offset - self.ext_dict_len) as _);
+            self.src_stream_offset = self.ext_dict_len;
```
It depends, after reposition some parts of the table may be reset to 0 which will be an invalid match or (more likely?) be a match that is too far back.

PSeitz commented:
Thank you for your patience, really exceptional quality. One thing that's missing I think are tests for the features exposed via `FrameInfo`, content_size, block_size and the two checksums. Currently only defaults are tested?

PSeitz commented:
> Related to this. Can you enable the github actions for this PR? It says
> 
> ```
> 1 workflow awaiting approval
> First-time contributors need a maintainer to approve running workflows.
> ```

It's quite annoying, I constantly need to click on approve and run, until one PR is merged.

> After a contributor has at least one pull request merged into a project's repository, any future pull requests from that contributor's fork will automatically run workflows.

arthurprs commented:
Thank you for nudging me about the tests, there were some horrendous bugs :facepalm: :facepalm: :facepalm: :facepalm: :facepalm:

PSeitz commented:
> Thank you for nudging me about the tests, there were some horrendous bugs facepalm facepalm facepalm facepalm facepalm

anything that is not covered by tests doesn't work or will be broken eventually :)

PSeitz commented on src/block/hashtable.rs:
```diff
@@ -197,14 +197,16 @@ impl HashTable for HashTableU16 {
-        16_001..=30_000 => (8192, 3),
-        // 100_000..=400_000 => (8192, 3),
-        _ => (16384, 2),
+        16_001..=65535 => (8192, 3),
+        // U32 positions => 16KB table
+        // Usize (U64) positions => 32KB table
+        _ => (4096, 4),
```
the hash table shrinks for sizes >65K

arthurprs replied:
It's intentional (as per commit msg), these were much larger then the reference implementation which is  16k bytes total.

PSeitz replied:
I mean the hashtable is larger for input sizes 16_001..=65535 than input sizes >65535.

In the block case for very small data (1KB) it was faster with a small table size if I remember correctly. I'm not sure how allocating a hashtable with 4k entries is faster than one with 256.

This also affects compression ratio, did you measure it?

PSeitz replied:
Ah I just saw the size wasn't forwarded anyway

arthurprs replied:
> This also affects compression ratio, did you measure it?

It does, yes. I don't have the numbers anymore though.

> Ah I just saw the size wasn't forwarded anyway

Exactly. So the U32 table was being 64KB total.

It should be possible to make the table size customizeable, but since the reference implementation never did (except in compile time) I doubt it's worth the hassle :shrug:

PSeitz replied:
Still, having a smaller hashtable for larger data doesn't make sense imo

```
16_001..=65535 => (8192, 3),
_ => (4096, 4),
```

arthurprs replied:
I see your point. Do you know how the previous numbers chosen? And what do you think it should ultimately look like?

I imagine the smaller tables for smaller inputs resulted in a speed up since it was faster to fill with zeroes, at little compression ratio cost. But if we're not sure I'd ratter settle for a behavior similar to the reference implementation which should have more empirical data behind it.

Either way this is only for the bloc compressor, which always knows the size of the input. The FrameCompressor, where matters the most, can continue to use the fixed 16KB table (as per reference).

arthurprs replied:
I suggest something like this, which is still arbitrary but still grows towards the 16KB limit faster.

```rust
pub fn get_table_size(input_len: usize) -> (usize, usize) {
    let (dict_size, dict_bitshift) = match input_len {
        // U16 Positions
        0..=65535 => {
            // A table up to 16K bytes in size, considering each slot takes 2 bytes.
            let size = input_len.next_power_of_two().min(16 * 1024) / 2;
            (size, 16 - size.trailing_zeros() as usize)
        }
        // U32 positions => 16KB table
        // Usize (U64) positions => 32KB table
        _ => (4096, 4),
    };
    (dict_size, dict_bitshift)
}

```

PSeitz replied:
I did some measurements on speed and compression and interpolated some values.
The scenario would be a service which receives compressed data of mostly small data. It would be wasteful to allocate 16kb for e.g. 1kb packets.

Such a function would be fine for me, but we should test for all sizes, that we get correct pairs. When the bitshift is wrong, we would waste size in the hashmap or it would fail.

arthurprs replied:
Adjusted in d197a09

PSeitz commented on src/frame/decompress.rs:
```diff
@@ -399,3 +398,31 @@ impl<R: fmt::Debug + io::Read> fmt::Debug for FrameDecoder<R> {
     }
 }
+
+/// Similar to set_len but panics if the vec doesn't have enough capacity.
```
this method doesn't panic

arthurprs replied:
The unsafe version does :thinking:  Should l replace the debug_assert with a plain assert?

PSeitz replied:
I would just adjust the comments

PSeitz commented on src/frame/decompress.rs:
```diff
@@ -399,3 +398,31 @@ impl<R: fmt::Debug + io::Read> fmt::Debug for FrameDecoder<R> {
+/// Similar to `v.get_mut(start..end) but will adjust the len if needed.
+/// Panics if there's not enough capacity.
+#[inline]
+fn vec_bytes_mut(v: &mut Vec<u8>, start: usize, end: usize) -> &mut [u8] {
```
`resize_and_get_mut` would be a better name I think

arthurprs replied:
Indeed, thanks.

PSeitz commented on src/block/hashtable.rs:
```diff
@@ -198,15 +198,26 @@ impl HashTable for HashTableU16 {
-        8_001..=16_000 => (2048, 5),
-        16_001..=65535 => (8192, 3),
+        0..=65535 => {
+            // Considering we want a table with up to 16K bytes and each slot takes 2 bytes.
```
Do you mean with each slot, a slot in the hash table?
Because 2 bytes is the case for inputs < U16::MAX, otherwise a slot takes 4bytes.

Therefore, in terms of byte size there is a peak for input sizes between 16384 and 32768.

```
16384 -> 8192, 3
32768 -> 8192, 3
65536 -> 4096, 4
131072 -> 4096, 4
```

PSeitz commented:
Time to merge finally :D
Really great job!

arthurprs commented:
Thank you :tada: Looking forwarding to using this very soon.

PSeitz commented on src/block/decompress.rs:
```diff
@@ -1,43 +1,91 @@
+    // Considering that `wild_copy_match_16` can copy up to `16 - 1` extra bytes.
+    // Defer to `duplicate_overlapping` in case of an overlapping match
+    // OR the if the wild copy would copy beyond the end of the output.
+    if start.add(match_length + 16 - 1) > *output_ptr
+        || output_ptr.add(match_length + 16 - 1) > output_end
```
I'm currently checking a performance regression for unsafe decompression (66K JSON 5GB/s down to 4GB/s). It is mainly caused by this extended check

arthurprs replied:
That's odd, 1Gbp/s from this branch in the slow path? The compiler is going crazy.

arthurprs replied:
This commit comes to mind https://github.com/PSeitz/lz4_flex/pull/13/commits/0a158955e4ff742a5b494c23d87a3d397c1661c4

But it's in a different location.

arthurprs replied:
Something that can help a lot is to run both versions alongside, in same binary.
The bigger benchmark binaries now can throw LTO offtrack significantly, it's almost random sometimes.

```toml
[dev-dependencies.lz4_flex_master]
git = "https://github.com/PSeitz/lz4_flex"
rev= "eb8c2e090485dcb5ef9e0da96bf72a95023753c1" # before this PR was merged
package = "lz4_flex"
```

Then it's available as `lz4_flex_master` in the benchmark code.

arthurprs replied:
I tried the above and this branch is 15% faster so something might be affecting your comparison.

```
BlockDecompress/lz4_flex_rust_master/66675                                                                             
                        time:   [10.145 us 10.153 us 10.162 us]
                        thrpt:  [6.1108 GiB/s 6.1161 GiB/s 6.1208 GiB/s]
                 change:
                        time:   [-2.2150% -2.1219% -2.0144%] (p = 0.00 < 0.05)
                        thrpt:  [+2.0558% +2.1679% +2.2652%]
                        Performance has improved.
Found 2 outliers among 100 measurements (2.00%)
  2 (2.00%) high mild
```
```
BlockDecompress/lz4_flex_rust/66675                                                                             
                        time:   [8.7167 us 8.7213 us 8.7266 us]
                        thrpt:  [7.1157 GiB/s 7.1200 GiB/s 7.1238 GiB/s]
                 change:
                        time:   [-2.0332% -1.6904% -1.3140%] (p = 0.00 < 0.05)
                        thrpt:  [+1.3315% +1.7195% +2.0754%]
                        Performance has improved.
```

PSeitz replied:
Hmm, I had 5Gb/s per second before on master, but yes flex_master (3,6GB/s) is slower than the current.
I'm not sure that flex_master is the same as having the crate locally with regards to inlining?
I changed the check to:
```   
if start as usize + match_length + 16 - 1 > *output_ptr as usize
{
```
which led to a speedup from 4Gb/s to 4,7Gb/s, which is weird, it should not have that much of an effect. I changed some of the benchmark code and cannot reproduce it anymore (it's 4,4Gb/s now). So I guess it is related to some LTO stuff in the benchmark code.

Thanks for double checking.

arthurprs replied:
Yeah I find it very disturbing that the performance can vary so much due to  seemingly unrelated stuff 😩

PSeitz replied:
https://www.youtube.com/watch?v=r-TLSBdHe1A&t=642s this talk is addresses this issue of random memory layout changes affecting performance

PSeitz commented:
released with 0.8.0
